File: src/levels/L06-decorator/decorator_caramelo.rb
require "delegate"

# region: Latte
class Latte
  def cost = 45
  def description = "Latte"
end
# endregion

# Cada extra envuelve a la bebida y suma lo suyo. Se combinan solos.
# region: WithOatMilk
class WithOatMilk < SimpleDelegator
  def cost = super + 8
  def description = "#{super} con avena"
end
# endregion

# region: WithShot
class WithShot < SimpleDelegator
  def cost = super + 10
  def description = "#{super} con shot"
end
# endregion

# region: WithCinnamon
class WithCinnamon < SimpleDelegator
  def cost = super + 5
  def description = "#{super} con canela"
end
# endregion

# region: WithCaramel
class WithCaramel < SimpleDelegator # ← un extra nuevo: una clase, nada más
  def cost = super + 7
  def description = "#{super} con caramelo"
end
# endregion

class Menu
  EXTRAS = { avena: WithOatMilk, shot: WithShot, canela: WithCinnamon, caramelo: WithCaramel }

  # region: Menu#drink_for
  def drink_for(request)
    request.extras.reduce(Latte.new) { |drink, extra| EXTRAS.fetch(extra).new(drink) }
  end
  # endregion
end

File: src/levels/L06-decorator/ts/decorator_caramelo.ts
// region: Latte
class Latte implements Drink {
  cost() { return 45 }
  description() { return 'Latte' }
}
// endregion

// Cada extra envuelve a la bebida y suma lo suyo. Se combinan solos.
// region: WithOatMilk
class WithOatMilk implements Drink {
  constructor(private drink: Drink) {}
  cost() { return this.drink.cost() + 8 }
  description() { return `${this.drink.description()} con avena` }
}
// endregion

// region: WithShot
class WithShot implements Drink {
  constructor(private drink: Drink) {}
  cost() { return this.drink.cost() + 10 }
  description() { return `${this.drink.description()} con shot` }
}
// endregion

// region: WithCinnamon
class WithCinnamon implements Drink {
  constructor(private drink: Drink) {}
  cost() { return this.drink.cost() + 5 }
  description() { return `${this.drink.description()} con canela` }
}
// endregion

// region: WithCaramel
class WithCaramel implements Drink { // ← un extra nuevo: una clase, nada más
  constructor(private drink: Drink) {}
  cost() { return this.drink.cost() + 7 }
  description() { return `${this.drink.description()} con caramelo` }
}
// endregion

class Menu {
  static EXTRAS: Record<string, new (drink: Drink) => Drink> = { avena: WithOatMilk, shot: WithShot, canela: WithCinnamon, caramelo: WithCaramel }

  // region: Menu#drink_for
  drinkFor(request: DrinkRequest): Drink {
    return request.extras.reduce<Drink>((drink, extra) => new Menu.EXTRAS[extra](drink), new Latte())
  }
  // endregion
}
